# WEEK 1
# 2. Simple arithmetical and logical expressions in Julia

# Arithmetical Expressions:

<hr>
<h2>Outcome</h2>

After this lecture, you will be able to: 

- Form arithmetical expressions using numbers and operators
- State the order of preference of Julia's elementary arithmetical operators
- Work out exactly how Julia would evalutate an arithmetical expression

# Simple example on arithmetical expressions: 

In [106]:
1-2+3-4

-2

In [107]:
1/ 2+3 /4

1.25

In [108]:
1/ (2+3) /4

0.05

In [109]:
1/ (2+3 /4)

0.36363636363636365

In [110]:
(1/2+3) /4

0.875

Operator Precedence in Julia:

The Plus(+) , Minus(-), Multiply(*) and Divide(/) are the operators in Julia.
When we have several of them then of occur it can be ambiguous, and then we can resolve the ambiguity in a number of different ways by putting parentheses in there. But Julia doesn't know 
exactly where we want to put the parentheses.
So, in Julia, when we have several events so 1/2/3. So what Julia does is it simply goes from left to right. So you have 1/2 is a half and then you divide that by 3, and that should of course be 1/6 and from the right to it will give 3/2



In [111]:
println(1/2/3)

0.16666666666666666


In [112]:
println(1/6)

0.16666666666666666


In [113]:
println(2-3+4)

3


In [114]:
println(2/3^2)

0.2222222222222222


So what is going on here? So although the division is the first operator, it's to the left of the power. The latter is evaluated first, and this is because some operators come in a higher order of preference and others.

In [115]:
println(2-3/4)

1.25


In [116]:
println(3+2*9-6)

15


So as in many other languages, particular Matlab and Python:

The order of the operations are, first, we do the insides of parenthesis. So if we have many parentheses, we do the inner parentheses before the outer ones. 

Then we do exponentiation. 

Then we do multiplication. We do that from left to right. 

Then we do addition and subtraction, from left to right.

<h3 id="int"> The exponentiation operator is actually different from the other four.</h3>

When several of them are stacked together, they operate from right to left.

In [117]:
println(3^2^3)    # exponentiation right to left is the julia convection

6561


In [118]:
println((3^2)^3)    #forcing left to right using parentheses

729


So this is the standard mathematical way, and many other computer languages also do a stack of exponentiation from right to left, and everything else from left to right. 

# Logical Expressions: 

Logical expressions are really what set computer languages apart from arithmetic. With logical expressions you get the computer language to do a whole variety of different things

<hr>
Outcome:

1.	Say what computer logic is,
2.	Explain the role of the values true and false in Julia. 
3.	You'll to be able to form logical expressions using not, and, and or.
4.  State the order of preference when you use Julia's arithmetical and logical operators.
5.  Work out exactly how Julia would evaluate a logical expression.



<h3 id="int">  The keywords true and false. </h3>

So, all of computer logic, at least as far as we're concerned in Julia, is about two logical values, the value true and the value false. They're reserved words. So you can only ever use them in the way that they have been reserved for. And that is that true indicates what is logically true and that false indicates what is logically false.

# Logical Operators and Logical Expressions.

So, Julia has actually quite a few logical operators, but we will focus on these three.

1. ! is the NOT operator.
2. && is the AND operator.
3. || is the OR operator.

The Not (!) operator changes the truth of whatever it is applied to

In [119]:
!true      # the opposite of NOT true is of course false.

false

In [120]:
!false       # the opposite of NOT false is of course true.

true

The other two operators AND and OR, they combine two logical values. So, let's call these logical values p and q, p is either true or false, q is either true or false.

The AND operator then makes a compound statement, a statement that consists of both p and q, and it is true if and only if both parts are true. So, this true and true, that must come out is true. And all of the other possible combinations, false and true, true and false, false false.

In [121]:
true && true

true

In [122]:
false && true

false

In [123]:
true && false

false

In [124]:
false && false

false

The OR operator needs only one of the parts to be true.

The OR operator makes a compound statement that is false if and only both parts are false. So, if we start with true and true, then that's true. And if we start with false and true, that's true. And if we start with true and false, that's true. But if we have false OR false, then it can only be false.

In [125]:
true || true

true

In [126]:
true || false

true

In [127]:
false || true

true

In [128]:
false || false

false

Sometimes we say a thing like, either the train is late or the station clock is wrong. We would then hope to eliminate one of them by showing it is false so that the other one is what we go by. And for this to work, at least one of them must be true. And that's exactly what we do with the OR operator.

# Comparison Operators

So, here's a comparison operator, less than (<), bigger than (>), and equals (==).

In [129]:
1 < 22

true

In [130]:
-22 > 1

false

In [131]:
2 == 4-2

true

So, these operators occur all the time in programming and you have to be very careful that you understand them very well.

# Operator Precendence for Logical Operators

If you write the code, you will often have to construct logical expressions very similar to the example.

3-4 > 1 && 2+2  == 4 || 10-5 > 2

Julia however, makes sure that there is only one way to evaluate this expression. And by putting in parentheses in the correct places



In [132]:
(3-4 > 1) && (2+2  == 4) || (10-5 > 2)    # we first do the arithmetic and then we do the comparison

true

In [133]:
false && true || true # evaluation from left to right.

true

# Practice Quiz

<h2 id="int"> Question 1      </h2>

Using the Julia environment what is the result of executing the following line of code:

5040 / 60

a. 84.0

b. 48

c. 48.0

d. 0.011904761904761904

e. 84

Double-click __here__ for the solution.

<!-- Your answer is below:
5040 / 60
84.0

Answer: A
-->

<h2 id="int"> Question 2      </h2>

As with ordinary mathematical notation, Julia evaluates a series of exponentiation operators from right to left. 

For example, 2^2^3 is equivalent to 2^(2^3).

What is the exact value of 2^2^1^1.3^1.5^1.7^20?

Double-click __here__ for the solution.

<!-- Your answer is below:
2^2^1^1.3^1.5^1.7^20
4.0

Answer: 4.0
-->

<h2 id="int"> Question 3      </h2>

Julia allows one to omit the multiplication operator when writing expressions. Which of the following expressions is equivalent to:

3 * x^2-2 * x+3

a. 3x^2-2x+3

b. x3^2-2x+3

c. 3x^2-3+2*x

d. 3x.x-2x+3

Double-click __here__ for the solution.

<!-- Your answer is below:
x=1
3x^2-2x+3
4
3x^2-2x+3
4
Answer: A
-->

<h2 id="int"> Question 4      </h2>

Suppose we write the following logical expression in Julia:

(3.0 < x) && (x < 5.0)

Which of the following are true?


a. The expression evaluates as true for 

b. x equal to 4 and no other integers.

c. The expression evaluates as true for all real numbers 

d. x in the closed interval from 3 to 5. Julia represents real numbers using the float type.



Double-click __here__ for the solution.

<!-- Your answer is below:
x=4   
    
(3.0 < x) && (x < 5.0)
    
 True   
Answer: A
-->

<h2 id="int"> Question 5     </h2>

If the Julia variable a is set as follows a=true and afterwards b is set to b=!!!!a then what is the value of b?

a. true

b. false

d. error

Double-click __here__ for the solution.

<!-- Your answer is below:
a=true 
    
b=!!!!a
    
True   
    
Answer: A
-->

<h2 id="int"> Question 6      </h2>

Consider the following two logical expressions

##### First expression

3 & 5 > 0 

##### Second expression

8 & 5 > 0

When evaluated, what values are returned for each of these expressions?


a. The result is false and false respectively.

b. The result is true and true respectively.

c. The result is true and false respectively.

d. The result is false and true respectively.

    
    


Double-click __here__ for the solution.

<!-- Your answer is below:
3 & 5 > 0
true 
    
8 & 5 > 0
    
false   
    
Answer: C
-->

<h2 id="int"> Question 7     </h2>

As it stands, the expression

1 + 2 - 5 - 2 == 0 

is both false and ambiguous.

There is one way to add a pair of parentheses on the left to make it true and unambiguous. Which of the following is it?


a. 1 + 2 - (5 - 2) == 0


b. 1 + (2 - 3) - 2 == 0


c. (1 + 2 - 5) - 2 == 0



Double-click __here__ for the solution.

<!-- Your answer is below:
1 + 2 - 5 - 2 == 0
false 
    
1 + 2 - (5 - 2) == 0
true   
    
Answer: A
-->